### ISSUE TITLE
Offscreen problem in P2D/P3D in beta 7

### ISSUE THREAD

**Issue by tillnagel** [processing/processing#3858]
Apologies for this bad bug report. I have no clue what is happening and need help. Since beta 7 I have weird glitches when using Unfolding Maps, and would love to get some pointers on where to investigate.

When using Unfolding with P2D or P3D, sometimes the screen stays black, sometimes a smaller map appears, and (very seldomly) everything works. I did not change anything, and in beta 6 it works.

Again, no clue, all my testing with offscreen PGraphics work, and I am at lost how to proceed. Tried looking through what changed with OpenGL between b6 and b7, but did not find any leads. Unfolding uses two stacked offscreen buffers (when P2D/P3D), and lots of matrix calculations. Does any of this triggers anything, so I can further investigate?

**Comment by tillnagel** [comment]
Ok, at least I have a sketch now, showing _some_ very weird behavior. Might be unrelated. 

When running this sketch, after a while (20-30s) the screen changes: from default background (gray) to black, and the matrix translation is not being used at all times (i.e. the image is displayed twice, at intended position and at window origin). Funnily, the second offscreen buffer is only initiated but not used, and still when removed this does not happen.

```
PImage testTile;
PGraphics pg;
PGraphics pg2;

public void settings() {
  size(800, 600, P3D);
}

public void setup() {
  testTile = loadImage("http://a.basemaps.cartocdn.com/light_all/1/1/1.png");
  pg = createGraphics((int) width, (int) height, P3D);
  pg.smooth(g.smooth);
  pg2 = createGraphics((int) width, (int) height, P3D);
}

public void draw() {
  pg.beginDraw();
  pg.clear();
  pg.image(testTile, 0, 0);
  pg.endDraw();

  pushMatrix();
  translate(width/2, height/2);
  image(pg, 0, 0);
  popMatrix();

  fill(255, 0, 0);
  ellipse(mouseX, mouseY, 100, 100);
}

```

First, this results in expected screen:

![first](https://cloud.githubusercontent.com/assets/908847/10043280/0602edec-61f2-11e5-90c0-440dcdb7c458.png)

Then, after a while in this:

![second](https://cloud.githubusercontent.com/assets/908847/10043283/0ef8b8dc-61f2-11e5-91f4-ece320a5b108.png)

**Comment by codeanticode** [comment]
I can reproduce the strange behavior of your test sketch. It happens irrespectively of creating pg2 or not. I will look into it, wondering if it could be related to https://github.com/processing/processing/commit/3335f2e2d2d4c547ae43cf19bf7487ab010e76bd

**Comment by aradke** [comment]
I am seeing quite a bit of strange behavior in Beta 7 too for offscreen createGraphics with P2D.  

See the following script and results for different renders. Since the first frame seems to be drawn ok, some frame variation needs to take place to see the issue. 

```java
PGraphics pg;
String render = P2D;

void setup() {
  size(200, 200, render);
  pg = createGraphics(100, 100, render);
}

void draw() {
  pg.beginDraw();  pg.ellipse(50+20*cos(frameCount), 50+20*sin(frameCount), 4,4);  pg.endDraw();
  image(pg, 0, 0); 
  text(render,10,10);
}

```

**JAVA2D**: This looks as expected.  A circle should be drawn.
![java2d](https://cloud.githubusercontent.com/assets/14793/10052998/9945d42a-61f7-11e5-92eb-5c8394828c2c.png)

**P2D**: The first frame draws a little circle in the right place, but then the following frames contain a significant offset. _(Note, increasing the y-size of the sketch moves this offset further down)_ 
![p2d](https://cloud.githubusercontent.com/assets/14793/10052996/99441e46-61f7-11e5-9b09-0ccf024be8a8.png)

**P3D**: Not sure why nothing shows up.
![p3d](https://cloud.githubusercontent.com/assets/14793/10052997/99444df8-61f7-11e5-8d65-bd807193d35f.png)

**FX2D**: Fails altogether, but that is a different issue.

**Comment by codeanticode** [comment]
@aradke thanks for reporting this other issue. However, I'm not able to reproduce it, although I'm on OS X, while it seems you are on windows. Out of curiosity, are you able to reproduce the issue reported by @tillnagel?

**Comment by aradke** [comment]
@codeanticode I used the win-64bit version on Win7.  

I was able to reproduce the strange 2min delayed blackout from @tillnagel, however the map image didn't show up until the 2min blackout and only showed up in the upper left.  

I somehow think this this behavior is related to the issue I am seeing.  I originally thought the issue was the same sort if mirror image problem until I added some motion. A similar sketch just drawing an ellipse onto the graphics buffer yields:

```java
PGraphics pg;
String render = P2D;

void setup() {
  size(200, 200, render);
  pg = createGraphics(100, 100, render);
}

void draw() {
  pg.beginDraw();  pg.ellipse(20,20, 20,20);  pg.endDraw();
  image(pg, 0, 0); 
  text(render,10,10);
}
```

**JAVA2D** _single_ ellipse
![ellipse-java2d](https://cloud.githubusercontent.com/assets/14793/10056682/3ef98918-620a-11e5-97c8-583f7b391075.png)
**P2D** _mirrored_ ellipse
![ellipse-p2d](https://cloud.githubusercontent.com/assets/14793/10056683/3eff05a0-620a-11e5-8fe7-2a462934c36e.PNG)
**P3D** _missing_ ellipse
![p3d](https://cloud.githubusercontent.com/assets/14793/10056684/3f03b294-620a-11e5-9b9f-256e45945b18.png)

Maybe some timing issue is involved too..  If I add a `delay` value in there....

```java
PGraphics pg;
String render = P2D;
int delay= 1000; //400 fails, 500 is intermittent,  1000 is normal at first and at a timerange gets a mirror

void setup() {
  size(200, 200, render);
  pg = createGraphics(100, 100, render);
}

void draw() {
  for(int i=0; i < delay; i=i+1){rect(0,0,10,10);}
  pg.beginDraw();  pg.ellipse(20,20, 20,20);  pg.endDraw();
  image(pg, 0, 0);
  text(render,10,10);
}
```

 I get various results. that produce some of this delayed results that @tillnagel saw.
`delay=400` fails,
`delay=500` is intermittent,
`delay=1000` at first looks right (like the JAVA2D) but then after some delay the mirror image returns.

**Comment by codeanticode** [comment]
This is indeed is a regression introduced by https://github.com/processing/processing/commit/3335f2e2d2d4c547ae43cf19bf7487ab010e76bd. Working on a fix now.

**Comment by codeanticode** [comment]
This commit https://github.com/processing/processing/commit/707703260c5ce74fcd27c8af24631fc015912dfa should address the problem. I run all test sketches posted so far, and did not observe any glitches affecting offscreen rendering.

If you have the chance to build Processing from source and want to run more tests, please let me know if you find any issues.

**Comment by aradke** [comment]
Great, I will give it a try tomorrow along with some other tests.

**Comment by tillnagel** [comment]
After some troubles building Processing (due to #3795) I managed to create the core libs, add them to my Eclipse env and run some Unfolding examples. It works again. Thanks @codeanticode!

**Comment by tillnagel** [comment]
Also, my original sketch does not result in the strange behavior anymore.

**Comment by codeanticode** [comment]
great, thanks for checking!

**Comment by aradke** [comment]
Fantastic @codeanticode.  707703260c5ce74fcd27c8af24631fc015912dfa solves the problem in all three sample scripts I submitted.  I also verified the [one posted](https://github.com/processing/processing/issues/3858#issuecomment-142564740) by @tillnagel is working on win8-64bit. 

Additionally, (after dropping in the 707703260c5ce74fcd27c8af24631fc015912dfa core.jar) it solves the issues I was seeing in a much more complicated multi-overlayed buffer example where I originally saw the issue in win7-64bit. 

On-wards to 3.0!
